lls/cache.c

```diff
@@ -310,6 +311,76 @@ lls_process_mod(struct lls_config *lls_conf, struct lls_mod_req *mod_req)
 	}
 }
 
+static void
+xmit_icmpv6_reply(struct gatekeeper_if *iface, struct rte_mbuf *pkt)
+{
+	/*
+	 * The icmpv6 header offset in terms of the
+	 * beginning of the IPv6 header.
+	 */
+	int icmpv6_offset;
+	uint8_t nexthdr;
+	struct ether_addr eth_addr_tmp;
+	struct ether_hdr *icmp_eth = rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+	struct ipv6_hdr *icmp_ipv6;
+	struct icmpv6_hdr *icmpv6_hdr;
+	size_t l2_len = pkt_in_l2_hdr_len(pkt);
+
+	ether_addr_copy(&icmp_eth->s_addr, &eth_addr_tmp);
+	ether_addr_copy(&icmp_eth->d_addr, &icmp_eth->s_addr);
+	ether_addr_copy(&eth_addr_tmp, &icmp_eth->d_addr);
+	if (iface->vlan_insert)
+		fill_vlan_hdr(icmp_eth, iface->vlan_tag_be, ETHER_TYPE_IPv6);
+
+	/* Set-up IPv6 header. */
+	icmp_ipv6 = (struct ipv6_hdr *)pkt_out_skip_l2(iface, icmp_eth);
+
+	/*
+	 * The IP Hop Limit field must be 255 as required by
+	 * RFC 4861, sections 7.1.1 and 7.1.2.
+	 */
+	icmp_ipv6->hop_limits = 255;
+	/*
+	 * According to RFC 4443 section 4.2, the Destination Address of an
+	 * ICMPv6 Echo Reply message is copied from the Source Address field
+	 * of the invoking Echo Request packet.
+	 */
+	rte_memcpy(icmp_ipv6->dst_addr, icmp_ipv6->src_addr,
+		sizeof(icmp_ipv6->dst_addr));
+	/*
+	 * According to RFC 4443 section 4.2, in case an Echo Request message
+	 * was sent to an IPv6 multicast or anycast address,
+	 * the source address of the reply MUST be a unicast address
+	 * belonging to the interface on which the Echo Request message
+	 * was received.
+	 */
+	rte_memcpy(icmp_ipv6->src_addr, iface->ip6_addr.s6_addr,
+		sizeof(icmp_ipv6->src_addr));
+
+	/* Set-up ICMPv6 header. */
+	icmpv6_offset = ipv6_skip_exthdr(icmp_ipv6, pkt->data_len - l2_len,
+		&nexthdr);
+
+	RTE_VERIFY(icmpv6_offset >= 0 && nexthdr == IPPROTO_ICMPV6);
```

Comment:
We cannot use `RTE_VERIFY()` here because `ipv6_skip_exthdr()` may fail because the IPv6 header is wrong. That is, we'd be allowing a remote attacker to bring Gatekeeper down with a malformed header.

`xmit_icmpv6_reply()` must return an error and caller free the packet.